src/mscorlib/src/System/Runtime/Loader/AssemblyLoadContext.cs

```diff
@@ -49,38 +65,75 @@ public abstract class AssemblyLoadContext
         internal static extern void InternalStartProfile(string profile, IntPtr ptrNativeAssemblyLoadContext);
 #endif // FEATURE_MULTICOREJIT
 
-        protected AssemblyLoadContext()
+        static AssemblyLoadContext()
         {
-            // Initialize the ALC representing non-TPA LoadContext
-            InitializeLoadContext(false);
+            // We register the cleanup of all AssemblyLoadContext that have not been finalized in the AppDomain.ProcessExit
+            AppDomain.CurrentDomain.ProcessExit += OnProcessExit;
         }
 
-        internal AssemblyLoadContext(bool fRepresentsTPALoadContext)
+        protected AssemblyLoadContext() : this(false)
         {
-            // Initialize the ALC representing TPA LoadContext
-            InitializeLoadContext(fRepresentsTPALoadContext);
         }
-        
-        void InitializeLoadContext(bool fRepresentsTPALoadContext)
+
+        internal AssemblyLoadContext(bool fRepresentsTPALoadContext)
         {
-            // Initialize the VM side of AssemblyLoadContext if not already done.
-            GCHandle gchALC = GCHandle.Alloc(this);
-            IntPtr ptrALC = GCHandle.ToIntPtr(gchALC);
-            m_pNativeAssemblyLoadContext = InitializeAssemblyLoadContext(ptrALC, fRepresentsTPALoadContext);
+            // Add this instance to the list of alive ALC
+            lock (assemblyLoadContextAliveList)
+            {
+                if (isProcessExiting)
+                {
+                    // TODO: Add localized version
+                    throw new InvalidOperationException("Cannot create an AssemblyLoadContext when the process is exiting");
+                }
 
-            // Initialize event handlers to be null by default
-            Resolving = null;
-            Unloading = null;
+                // Initialize the VM side of AssemblyLoadContext if not already done.
+                isDefault = fRepresentsTPALoadContext;
+                var thisWeakHandle = GCHandle.Alloc(this, GCHandleType.Weak);
```

Comment:
GC has more work to do, JIT generates less efficient code in some cases, runtime data structures are bigger, ... .